### PR TITLE
Adds kernel computed field

### LINKED ISSUE #1246: Abstraction for using "custom" kernels to compute fields (plus an example)?
In a perfect world one would be able to use `AbstractOperations` for all output needs. But alas, there is #1241. The discussion on #1234 makes clear that this will continue to be a serious issue. There may also be output that cannot be expressed as an abstract operation.

One work around for advanced  users (which may be nice to have anyways) is to design an abstraction for `KernelComputedField`s that are computed based on a user-supplied kernel. I think usage could be something like (using [`ViscousDissipation`](https://github.com/CliMA/LESbrary.jl/blob/master/src/TurbulenceStatistics/viscous_dissipation.jl) as an example)

```julia
model = IncompressibleModel(...)

@kernel function compute_viscous_dissipation!(ϵ, grid, clock, νₑ, u, v, w)
    i, j, k = @index(Global, NTuple)

    Σˣˣ = ∂xᶜᵃᵃ(i, j, k, grid, u)
    Σʸʸ = ∂yᵃᶜᵃ(i, j, k, grid, v)
    Σᶻᶻ = ∂zᵃᵃᶜ(i, j, k, grid, w)

    Σˣʸ = (ℑxyᶜᶜᵃ(i, j, k, grid, ∂yᵃᶠᵃ, u) + ℑxyᶜᶜᵃ(i, j, k, grid, ∂xᶠᵃᵃ, v)) / 2
    Σˣᶻ = (ℑxzᶜᵃᶜ(i, j, k, grid, ∂zᵃᵃᶠ, u) + ℑxzᶜᵃᶜ(i, j, k, grid, ∂xᶠᵃᵃ, w)) / 2
    Σʸᶻ = (ℑyzᵃᶜᶜ(i, j, k, grid, ∂zᵃᵃᶠ, v) + ℑyzᵃᶜᶜ(i, j, k, grid, ∂yᵃᶠᵃ, w)) / 2

    @inbounds ϵ[i, j, k] = νₑ[i, j, k] * 2 * (Σˣˣ^2 + Σʸʸ^2 + Σᶻᶻ^2 + Σˣʸ^2 + Σˣᶻ^2 + Σʸᶻ^2)
end

νₑ = model.diffusivities.νₑ
u, v, w = model.velocities

viscous_dissipation = KernelComputedField(Cell, Cell, Cell, model, compute_viscous_dissipation;
                                          field_dependencies=(νₑ, u, v, w))
```

In this example `viscous_dissipation` is then a bonafide field with a fully-fledged `compute!` function that can be used in output, time-averaging, in an `AveragedField`, etc.

Users will need to write a kernel with `KernelAbstractions` language (which we'll have to document and provide examples for), but they won't have to write all the other boilerplate that goes into defining the struct and writing a `compute!` function (see again [`ViscousDissipation`](https://github.com/CliMA/LESbrary.jl/blob/master/src/TurbulenceStatistics/viscous_dissipation.jl) to get an idea of what this boiler plate looks like).

The kernel can have `field_dependencies`, which are passed into the `KernelComputedField` constructor. It also seems nice to pass `clock`.

### PR DESCRIPTION
**Pull request by tomchor**
Adds `KernelComputedField <: AbstractField` which is computed using a user-defined `@kernel` function via `KernelAbstractions.jl`. To do:

- [x] Implementation
- [x] Testing
- [x] Example / Docs
- [x] Test complex examples on GPUs

Closes #1246

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
I think GitHub lets us convert PRs to "draft PRs" which could be an alternative to putting "[WIP]" in the PR title.

image

**Comment by tomchor** [comment]
@glwagner Sorry for posting in the wrong place. Here's the same message again:

@glwagner I modified the docstring with a small example that I think illustrates the patterns relatively in this commit. I'm assuming that's what you meant by "example", right?

I've tested it already in some simple examples and it produces correct results. Is there anything else left besides testing this with an actual complex calculation that can only be done via KernelComputedField on a GPU?

**Comment by tomchor** [comment]
I already tested it for some complex calculations (TKE, Ri, Ro and Ertel PV) and the results are correct!

File: src/Fields/kernel_computed_field.jl [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    Example
+    =======
+
+    ```julia
+    using KernelAbstractions: @index, @kernel
+    using Oceananigans.Fields: AveragedField, KernelComputedField, compute!
+    using Oceananigans.Grids: Cell, Face
+    
+    @inline ψ²(i, j, k, grid, ψ, Ψ) = @inbounds (ψ[i, j, k] - Ψ[i, j, k])^2
+    @kernel function compute_var!(var, grid, ϕ, Φ)
```
Comment (ali-ramadhan) [inline src/Fields/kernel_computed_field.jl:25]: I know it's just an example but would `compute_perturbation!` be a better name?

**Review by ali-ramadhan — APPROVED** [review]
Looks great! I'm guessing this would simplify some of the statistics in LESbrary.jl?

File: src/Fields/kernel_computed_field.jl (hunk 1 above)
Comment (tomchor) [reply src/Fields/kernel_computed_field.jl:25]: I'm open to a change of names, but my idea was that it computes the variance, hence `compute_var!`. (Although in hindsight I should have just named it `compute_variance!` and be more explicit.

Comment (ali-ramadhan) [reply src/Fields/kernel_computed_field.jl:25]: Ah haha sorry I thought it was short for `compute_variable!`. Indeed `compute_variance!` would be more clearer.

Comment (tomchor) [reply src/Fields/kernel_computed_field.jl]: Done!

**Comment by ali-ramadhan** [comment]
@tomchor Just merged it! We should add you as a collaborator on the repo so you can merge your own PRs once they're approved.

**Pull request merged**